packages/arb-ts/src/lib/message/L2ToL1Message.ts

```diff
@@ -0,0 +1,374 @@
+    return this.getL2ToL1Events().map(log => {
+      const outboxAddr = this.getOutboxAddr(arbNetwork, log.batchNumber)
+
+      return L2ToL1Message.fromBatchNumber(
+        l1SignerOrProvider,
+        outboxAddr,
+        log.batchNumber,
+        log.indexInBatch
+      )
+    })
+  }
+}
+
+/**
+ * Conditional type for Signer or Provider. If T is of type Provider
+ * then L2ToL1MessageReaderOrWriter<T> will be of type L2ToL1MessageReader.
+ * If T is of type Signer then L2ToL1MessageReaderOrWriter<T> will be of
+ * type L2ToL1MessageWriter.
+ */
+export type L2ToL1MessageReaderOrWriter<
+  T extends SignerOrProvider
+> = T extends Provider ? L2ToL1MessageReader : L2ToL1MessageWriter
+
+export class L2ToL1Message {
+  public static fromBatchNumber<T extends SignerOrProvider>(
+    l1SignerOrProvider: T,
+    outboxAddress: string,
+    batchNumber: BigNumber,
+    indexInBatch: BigNumber
+  ): L2ToL1MessageReaderOrWriter<T>
+  public static fromBatchNumber<T extends SignerOrProvider>(
+    l1SignerOrProvider: T,
+    outboxAddress: string,
+    batchNumber: BigNumber,
+    indexInBatch: BigNumber
+  ): L2ToL1MessageReader | L2ToL1MessageWriter {
+    return l1SignerOrProvider instanceof Provider
+      ? new L2ToL1MessageReader(
+          l1SignerOrProvider,
+          outboxAddress,
+          batchNumber,
+          indexInBatch
+        )
+      : new L2ToL1MessageWriter(
+          l1SignerOrProvider,
+          outboxAddress,
+          batchNumber,
+          indexInBatch
+        )
+  }
+
+  // CHRIS: consider what else to move onto the base
+  public static async getL2ToL1MessageLogs(
+    l2Provider: Provider,
+    filter: Filter
+  ) {
+    const iface = ArbSys__factory.createInterface()
+    const event = iface.getEvent('L2ToL1Transaction')
+    const eventTopic = iface.getEventTopic(event)
+
+    const topics = filter.topics ? filter.topics : []
+    // CHRIS: keep the warn?
+    // if (!filter.fromBlock && !filter.toBlock)
+    // console.warn('Attempting to query from 0 to block latest')
+    const logs = await l2Provider.getLogs({
+      address: ARB_SYS_ADDRESS,
+      topics: [eventTopic, ...topics],
+      fromBlock: filter.fromBlock || 0,
+      toBlock: filter.toBlock || 'latest',
+    })
+
+    return logs.map(
+      log => iface.parseLog(log).args as unknown as L2ToL1EventResult
+    )
+  }
+
+  public static async getL2ToL1MessageLog(
+    l2Provider: Provider,
+    batchNumber: BigNumber,
+    indexInBatch: BigNumber
+  ) {
+    const batch = await this.getL2ToL1MessageLogs(l2Provider, {
+      topics: [null, null, hexZeroPad(batchNumber.toHexString(), 32)],
+    })
+
+    const indexItem = batch.filter(b => b.indexInBatch.eq(indexInBatch))
+    if (indexItem.length === 1) {
+      return indexItem[0]
+    } else if (indexItem.length > 1) {
+      // CHRIS: warn, error?
+    }
+
+    return null
+  }
+}
+
+export class L2ToL1MessageReader extends L2ToL1Message {
```

Comment:
Now that we have this nice abstraction, something like "l1BlocksUntilConfirmable" would be a cool (doesn't have to be in this PR)